src/core/core/rpc_engine.cpp

```diff
@@ -1030,21 +1024,19 @@ namespace dsn {
                 }
             }
 
-            // request is forwarded, we cannot use the original rpc session
+            // request is forwarded, we cannot use the original rpc session,
+            // so use client session to send response.
             else
             {
                 dbg_dassert(response->to_address.port() > MAX_CLIENT_PORT, 
                     "target address must have named port in this case");
 
-                auto sp = task_spec::get(response->local_rpc_code);
-                auto& hdr = *response->header;
-                auto& named_nets = _client_nets[sp->rpc_call_header_format];
-                network* net = named_nets[sp->rpc_call_channel];
-
-                dassert(nullptr != net, "network not present for rpc channel '%s' with format '%s' used by rpc %s",
-                    sp->rpc_call_channel.to_string(),
-                    sp->rpc_call_header_format.to_string(),
-                    hdr.rpc_name
+                // use the header format recorded in the message
+                network* net = _client_nets[response->hdr_format][RPC_CHANNEL_TCP];
```

Comment:
I think the cost will be small and it breaks further network providers plugged in the future. Anyway, let me worry about this in the next pr for quickly merging this one first.